### ISSUE TITLE
Boolean literals parsed as "0", "1"

### ISSUE THREAD

**Issue by dan-f** [linkeddata/rdflib.js#147]
Boolean literals are getting parsed as correctly having the boolean `datatype`, but (I argue) incorrectly having the value `"0"` or `"1"`, which I find to be a leaky abstraction.

I'd fix this myself, but I cannot understand how the n3 parser is implemented. I assume it's running some sort of DFA, but I'm finding the code difficult to understand.

One curiosity is that [the `BOOLEAN_DATATYPE` constant is not being used by the parser](https://github.com/linkeddata/rdflib.js/blob/master/src/n3parser.js#L202).

See attached screenshot for what I'm referring to.

![rdflib_boolean_literal](https://cloud.githubusercontent.com/assets/2146312/19491449/ffceffaa-9540-11e6-8067-a3e73aed7532.png)

**Comment by timbl** [comment]
The parser is a standard top-down recursive descent parser.   (It was written i python and transcoded using pyjs aka  pyjamas)

 Results are pushed onto a result stack [res] in the style of python parser history.   
Functions like commaSeparatedList = function(str, j, res, ofUris) parse the production in the grammar with their name.  Here str is the whole string to be parsed, j is the pointer, res an array to push result values onto, and ofUris just a flag to allow this code to cover two cases.
It returns a new value of the pointer or  < 0 if the string does not match the grammar.

The function nodeOrLiteral parses a node or or literal value syntax.
The first line 
```
    var j = this.node(str, i, res);
    if ((j >= 0)) {
        return j;
    }
```
tried named node syntax ( <fooo#bar>    ex:bar etc)  and if that fails t assumes literal.
In fact node() covers the true and false cases:

```
    var j = this.tok("true", str, i);
    if ((j >= 0)) {
        res.push(true);
        return j;
    }
    var j = this.tok("false", str, i);
    if ((j >= 0)) {
        res.push(false);
        return j;
    }
```
at line 847.  
Those raw boolean values are what get passed to makeStatement and store.add()
IndexedFormula.add() calls Node.fromValue() on each part of the quad whcih calls Literal.fromValue which ends up doing this in literal.js

```
  /**
   * @method fromBoolean
   * @static
   * @param value {Boolean}
   * @return {Literal}
   */
  static fromBoolean (value) {
    let strValue = value ? '1' : '0'
    return new Literal(strValue, null, XSD.boolean)
  }
```
at line 64.

But the issue isn't how to change the way it is stored -- it is the fact that for RDF purists all datatypes have a value WHICH IS A STRING.  That is connected with the fact that they are defined by XML datatypes spec
https://www.w3.org/TR/2004/REC-xmlschema-2-20041028/#boolean
where for XML it all about strings which can actually occur in the document.
(That spec allows either "0" and "1" or "true" and "false" but not true or false the boolean values.
The canonical representation for boolean is the set of literals {true, false}.)

So in an RDF/XML document you can use any of those strings ... but why does that dictate the way things are stored in the Literal object? 

Maybe though to keep the simple API compatible we should just make a toJS() method which returns a boolean from a boolean and a Date from a date and Number from a number.

**Comment by dan-f** [comment]
Thank you @timbl for the detailed answer!

> So in an RDF/XML document you can use any of those strings ... but why does that dictate the way things are stored in the Literal object?

This is precisely the question I'm hoping to answer :) IMHO RDF/XML's choice to represent the values `true` and `false` as `"1"` and `"0"` is a design decision which I don't believe should leak through into RDFJS space. As long as we can serialize a `Literal` node to RDF/XML, the type of `literal.value` needn't always be a string.

That said, perhaps some folks are depending on a literal's `value` property to be the exact string that originally occurred in the RDF/XML. Does anyone know if this is true?

My proposal is to allow `Literal` nodes to take on non-string values (integers, floats, booleans, dates, etc) as long as they can be properly serialized back to the various RDF formats. However I'd also settle for your suggestion of a `toJS()` or `toNative` method :smile:

**Comment by HolgerKnublauch** [comment]
Yes if .value is supposed to be the interoperable mechanism to query nodes then .value should IMHO return "true" and "false" (strings) for booleans. I ran into this problem. 0 and 1 sound indeed like a leaked optimization. Maybe a better design for the task force interfaces would be to only define functions such as .getLex() or .lex() which return strings only, while internal fields such as .value may be used for custom optimizations. This would more cleanly separate the storage and optimization requirements from the interchange interfaces, which should be 100% RDF compliant.

**Comment by elf-pavlik** [comment]
@HolgerKnublauch can you imagine to propose it in Issue or even PR to https://github.com/rdfjs/representation-task-force

**Comment by HolgerKnublauch** [comment]
At this stage I am not sure how much of this has already been discussed by the JS task force, because I have just recently looked into this topic. I have written my opinion above, and do believe that functions would be better than attributes. If the task force finds this input useful, it may take it up. Otherwise, I don't want to reopen topics that were already discussed ad nauseam.

**Comment by HolgerKnublauch** [comment]
I just noticed that since 2015 JavaScript now supports getter functions and these can basically be used instead of functions so that the syntax would not look different, assuming that implementations don't use .value to expose their internal representation.

**Comment by dan-f** [comment]
I think that part of the issue here is that the `Literal`/`NamedNode` data types are used as both low-level representations of parsed RDF data and, currently, as part of the higher-level graph interface. Something needs to sit between the (`'0'`, `'boolean'`) tuple and consumer code. Maybe it's implemented at the `Literal` level, but it could also be implemented through another interface. For the sake of of interoperability, I can see why we'd want to keep the `Literal` class the way it is, and mediate data access through another interface.
